File: sdks/js/test/live.test.ts
// BILLED acceptance probes. Every test here reaches a real gateway, a real
// provider and a real credit balance.
//
// # Why the gate is node:test's own `skip` and not an early `return`
//
// An early `return` inside the test body is a PASS to the runner, so a machine
// with no key, no gateway and no credits would print these probes as passing —
// release evidence that cannot distinguish a probe that ran from one that never
// executed a line. `{ skip }` is decided BEFORE the body runs, so the TAP output
// says `# SKIP` with its reason. Non-execution can be skipped or a failure; it
// can never be a pass.
//
//   npm test                       # every probe reports SKIP
//   NROUTER_LIVE=1 npm test        # runs them; needs the env below
//
// # And why a missing per-wire variable is a THROW, not a skip
//
// Once NROUTER_LIVE=1 is set the caller has asked for the billed probes. A
// missing variable at that point is a misconfigured live run, not a reason to
// report success — so `required()` throws and names the variable.
//
// # The route-family matrix
//
// Claude-through-/v1/messages was the only live acceptance here, so the wires
// customers actually reported broken — OpenAI chat completions, /v1/responses,
// and an opaque alias whose provider is not inferable from its name — were
// outside live evidence entirely. They are separate tests with separate model
// variables because a model is servable on the wires ITS provider declares and
// no others: one model cannot certify the matrix, and a single test that tried
// would fail for a reason that is not a defect.

const { test } = require('node:test');
const assert = require('node:assert/strict');

const { nRouter } = require('../dist/index');

const skip =
  process.env.NROUTER_LIVE === '1'
    ? false
    : 'billed: set NROUTER_LIVE=1, NROUTER_API_KEY and the per-wire model variables';

/**
 * The value of `name`, or a throw naming it. Reached only under
 * NROUTER_LIVE=1, where the caller has already asked for a billed run.
 */
function required(name: string): string {
  const value = process.env[name];
  if (!value) {
    throw new Error(
      `${name} is required for a live probe. Set NROUTER_LIVE=1, NROUTER_API_KEY, ` +
        'and the per-wire model variables, then run `NROUTER_LIVE=1 npm test`.',
    );
  }
  return value;
}

/** A client pointed at the gateway under test. */
function liveClient() {
  return new nRouter({
    apiKey: process.env.NROUTER_API_KEY,
    baseURL: process.env.NROUTER_BASE_URL ?? 'http://127.0.0.1:4000/v1',
    maxRetries: 0,
  });
}

/**
 * Every response must carry `x-nr-request-id`: it is the only handle a customer
 * has at support, and the join key for the spend row this call just wrote.
 */
function assertCorrelatable(meta: any, wire: string) {
  assert.ok(meta.requestId, `${wire} answered without x-nr-request-id`);
}

/**
 * Unpriced is not free (Rule #28, gateway gate 3). `x-nr-request-cost` is
 * ABSENT when the model is unpriced — never a zero — so the honest states are
 * exactly two, and a reported cost of 0 is a defect on either.
 */
function assertHonestCost(meta: any, wire: string) {
  if (meta.costStatus === undefined || meta.costStatus === null) return;
  assert.ok(
    meta.costStatus === 'exact' || meta.costStatus === 'unpriced',
    `${wire} reported x-nr-cost-status ${meta.costStatus}`,
  );
  if (meta.costStatus === 'exact') {
    assert.ok(
      meta.cost !== null && meta.cost !== undefined && meta.cost > 0,
      `${wire} claimed an exact cost that was not above zero`,
    );
  } else {
    assert.equal(meta.cost, null, `${wire} priced an unpriced response`);
  }
}

/**
 * The `/v1` paths `GET /v1/models` says this alias can be called on.
 *
 * The gateway renders `nrouter_endpoints` from the provider's own endpoint
 * declaration, so this is the discovery answer an SDK is supposed to use
 * instead of guessing a wire from the model name.
 */
function advertisedEndpoints(catalogue: any, model: string): string[] {
  assert.ok(
    Array.isArray(catalogue.data) && catalogue.data.length > 0,
    'GET /v1/models returned an empty catalogue',
  );
  const entry = catalogue.data.find((item: any) => item.id === model);
  assert.ok(entry, `${model} is not in this key's catalogue`);
  assert.ok(
    Array.isArray(entry.nrouter_endpoints),
    `${model} carries no nrouter_endpoints`,
  );
  return entry.nrouter_endpoints;
}

/** A raw POST plus the `x-nr-*` metadata the gateway reported for it. */
async function post(client: any, path: string, body: Record<string, unknown>) {
  const res = await client.nr.request(path, body);
  assert.equal(res.status, 200, `${path} answered ${res.status}: ${res.text}`);
  return { body: JSON.parse(res.text), meta: client.nr.meta(res.headers) };
}

test('live Claude Messages request returns billing metadata', { skip }, async () => {
  const client = liveClient();
  const response = await client.nr.messages({
    model: process.env.NROUTER_LIVE_MESSAGES_MODEL ?? 'claude-3-5-haiku-20241022',
    max_tokens: 2,
    messages: [{ role: 'user', content: 'Reply OK' }],
  });

  assert.ok(Array.isArray(response.body.content));
  assertCorrelatable(response.meta, '/v1/messages');
  assert.equal(response.meta.costStatus, 'exact');
  assert.ok(response.meta.cost !== null && response.meta.cost > 0);
});

test('live OpenAI chat completions wire answers', { skip }, async () => {
  const client = liveClient();
  const { body, meta } = await post(client, '/chat/completions', {
    model: required('NROUTER_LIVE_CHAT_MODEL'),
    max_tokens: 2,
    messages: [{ role: 'user', content: 'Reply OK' }],
  });

  assert.ok(Array.isArray(body.choices), '/v1/chat/completions returned no choices array');
  assertCorrelatable(meta, '/v1/chat/completions');
  assertHonestCost(meta, '/v1/chat/completions');
});

test('live Responses wire answers', { skip }, async () => {
  const client = liveClient();
  const response = await client.nr.responses({
    model: required('NROUTER_LIVE_RESPONSES_MODEL'),
    input: 'Reply OK',
    max_output_tokens: 16,
  });

  assert.ok(
    response.body && Object.keys(response.body).length > 0,
    '/v1/responses returned an empty document',
  );
  assertCorrelatable(response.meta, '/v1/responses');
  assertHonestCost(response.meta, '/v1/responses');
});

// An alias whose provider a client cannot infer from the name — a Bedrock GLM
// or a Gemma alias — must still be callable, and the wire must come from
// discovery rather than from a guess.
//
// This is the one probe that proves the matrix is DERIVABLE: it reads the
// endpoints out of GET /v1/models and then calls the wire it was told about.
// An alias listed with an endpoint it cannot serve fails here.
test('live opaque alias is callable on the wire discovery advertises', { skip }, async () => {
  const client = liveClient();
  const model = required('NROUTER_LIVE_OPAQUE_MODEL');
  const endpoints = advertisedEndpoints(await client.nr.models.list(), model);
  assert.ok(
    endpoints.length > 0,
    `${model} is listed with an empty nrouter_endpoints — the catalogue advertises a name no wire serves`,
  );

  const path = endpoints.includes('/v1/chat/completions')
    ? '/chat/completions'
    : endpoints.includes('/v1/messages')
      ? '/messages'
      : null;
  assert.ok(path, `${model} advertises no text wire: ${JSON.stringify(endpoints)}`);

  const { meta } = await post(client, path as string, {
    model,
    max_tokens: 2,
    messages: [{ role: 'user', content: 'Reply OK' }],
  });
  assertCorrelatable(meta, 'the discovered text wire');
  assertHonestCost(meta, 'the discovered text wire');
});

File: sdks/java/src/test/java/ai/nrouter/sdk/LiveTest.java
package ai.nrouter.sdk;

import static org.junit.jupiter.api.Assertions.*;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;

/**
 * BILLED acceptance probes. Every test here reaches a real gateway, a real
 * provider and a real credit balance.
 *
 * <p>The gate is {@link EnabledIfEnvironmentVariable}, decided BEFORE the body
 * runs, so an unconfigured machine reports these as <em>disabled</em> in the
 * surefire XML a release reads. An early {@code return} would have reported the
 * same machine as passing — evidence that cannot distinguish a probe that ran
 * from one that never executed a line. Non-execution can be disabled or a
 * failure; it can never be a pass.
 *
 * <p>Once {@code NROUTER_LIVE=1} is set the caller has asked for the billed
 * probes, so a missing per-wire variable is a misconfigured live run rather
 * than a reason to report success: {@link #require} FAILS and names it.
 *
 * <h2>The route-family matrix</h2>
 *
 * <p>Claude-through-{@code /v1/messages} was the only live acceptance here, so
 * the wires customers actually reported broken — OpenAI chat completions,
 * {@code /v1/responses}, and an opaque alias whose provider is not inferable
 * from its name — were outside live evidence entirely. They are separate tests
 * with separate model variables because a model is servable on the wires ITS
 * provider declares and no others.
 *
 * <p>The opaque-alias probe names its wire in {@code NROUTER_LIVE_OPAQUE_WIRE}
 * rather than reading {@code nrouter_endpoints} out of {@code GET /v1/models},
 * as the Rust, Kotlin and JS probes do. {@link NRouterHttpClient} exposes no
 * GET, and adding one to the shipping surface is a wider change than a test may
 * make. This probe therefore proves the alias is CALLABLE; it does not prove
 * discovery advertises the wire it was called on.
 */
@EnabledIfEnvironmentVariable(
        named = "NROUTER_LIVE",
        matches = "1",
        disabledReason = "billed: set NROUTER_LIVE=1, NROUTER_API_KEY and the per-wire model variables")
class LiveTest {
    /** The value of {@code name}, or a failure naming it. */
    private static String require(String name) {
        String value = System.getenv(name);
        assertNotNull(
                value,
                name + " is required for a live probe. Set NROUTER_LIVE=1, NROUTER_API_KEY, and the"
                        + " per-wire model variables, then run `NROUTER_LIVE=1 mvn test`.");
        return value;
    }

    /** A client pointed at the gateway under test. */
    private static NRouterHttpClient liveClient() {
        String apiKey = require("NROUTER_API_KEY");
        String baseUrl = System.getenv().getOrDefault("NROUTER_BASE_URL", "http://127.0.0.1:4000/v1");
        return NRouter.httpClient(apiKey, baseUrl);
    }

    /**
     * Every response must carry {@code x-nr-request-id}: it is the only handle a
     * customer has at support, and the join key for the spend row this call just
     * wrote.
     */
    private static void assertCorrelatable(NRouterHttpResponse response, String wire) {
        assertEquals(200, response.statusCode(), wire + " did not answer 200");
        assertNotNull(response.meta().requestId(), wire + " answered without x-nr-request-id");
        assertFalse(response.meta().requestId().isEmpty(), wire + " answered with an empty x-nr-request-id");
    }

    /**
     * Unpriced is not free (Rule #28). {@code x-nr-request-cost} is ABSENT when
     * the model is unpriced — never a zero — so the honest states are exactly
     * two, and a reported cost of 0 is a defect on either.
     */
    private static void assertHonestCost(NRouterHttpResponse response, String wire) {
        String status = response.meta().costStatus();
        if (status == null) {
            return;
        }
        assertTrue(
                "exact".equals(status) || "unpriced".equals(status),
                wire + " reported x-nr-cost-status " + status);
        if ("exact".equals(status)) {
            assertNotNull(response.meta().cost(), wire + " claimed an exact cost and sent none");
            assertTrue(response.meta().cost() > 0.0, wire + " claimed an exact cost that was not above zero");
        } else {
            assertNull(response.meta().cost(), wire + " priced an unpriced response");
        }
    }

    private static Map<String, Object> textBody(String model) {
        return Map.of(
                "model", model,
                "max_tokens", 2,
                "messages", List.of(Map.of("role", "user", "content", "Reply OK")));
    }

    @Test
    void liveClaudeRequestReturnsBillingMetadata() {
        String model = System.getenv().getOrDefault("NROUTER_LIVE_MESSAGES_MODEL", "claude-3-5-haiku-20241022");
        NRouterHttpResponse response = liveClient().messages(textBody(model));

        assertCorrelatable(response, "/v1/messages");
        assertTrue(response.meta().isPriced(), "/v1/messages did not price a priced model");
        assertTrue(response.meta().cost() > 0.0);
    }

    @Test
    void liveOpenAiChatCompletionsWireAnswers() {
        NRouterHttpResponse response = liveClient().chatCompletions(textBody(require("NROUTER_LIVE_CHAT_MODEL")));

        assertTrue(
                response.body().path("choices").isArray(),
                "/v1/chat/completions returned no choices array");
        assertCorrelatable(response, "/v1/chat/completions");
        assertHonestCost(response, "/v1/chat/completions");
    }

    @Test
    void liveResponsesWireAnswers() {
        NRouterHttpResponse response = liveClient()
                .responses(Map.of(
                        "model", require("NROUTER_LIVE_RESPONSES_MODEL"),
                        "input", "Reply OK",
                        "max_output_tokens", 16));

        assertFalse(response.body().isMissingNode(), "/v1/responses returned an empty document");
        assertCorrelatable(response, "/v1/responses");
        assertHonestCost(response, "/v1/responses");
    }

    /**
     * An alias whose provider a client cannot infer from the name — a Bedrock
     * GLM or a Gemma alias — must still be callable. See the class header for
     * why this probe is told its wire instead of discovering it.
     */
    @Test
    void liveOpaqueAliasIsCallable() {
        String model = require("NROUTER_LIVE_OPAQUE_MODEL");
        String wire = System.getenv().getOrDefault("NROUTER_LIVE_OPAQUE_WIRE", "/v1/chat/completions");
        NRouterHttpClient client = liveClient();
        // if/else, not a switch expression: this module targets Java 11.
        NRouterHttpResponse response;
        if ("/v1/chat/completions".equals(wire)) {
            response = client.chatCompletions(textBody(model));
        } else if ("/v1/messages".equals(wire)) {
            response = client.messages(textBody(model));
        } else {
            throw new IllegalStateException(
                    "NROUTER_LIVE_OPAQUE_WIRE must be /v1/chat/completions or /v1/messages, not " + wire);
        }

        assertCorrelatable(response, wire);
        assertHonestCost(response, wire);
    }
}
